core/src/storage2/lazy/lazy_cell.rs

```diff
@@ -0,0 +1,421 @@
+{
+    /// The key to lazily load the value from.
+    ///
+    /// # Note
+    ///
+    /// This can be `None` on contract initialization where a `LazyCell` is
+    /// normally initialized given a concrete value.
+    key: Option<Key>,
+    /// The low-level cache for the lazily loaded storage value.
+    ///
+    /// # Safety (Dev)
+    ///
+    /// We use `UnsafeCell` instead of `RefCell` because
+    /// the intended use-case is to hand out references (`&` and `&mut`)
+    /// to the callers of `Lazy`. This cannot be done without `unsafe`
+    /// code even with `RefCell`. Also `RefCell` has a larger memory footprint
+    /// and has additional overhead that we can avoid by the interface
+    /// and the fact that ink! code is always run single-threaded.
+    /// Being efficient is important here because this is intended to be
+    /// a low-level primitive with lots of dependencies.
+    cache: UnsafeCell<Option<Entry<T>>>,
+}
+
+impl<T> Debug for LazyCell<T>
+where
+    T: Debug + SpreadLayout,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("LazyCell")
+            .field("key", &self.key)
+            .field("cache", unsafe { &*self.cache.get() })
+            .finish()
+    }
+}
+
+#[test]
+fn debug_impl_works() {
+    let c1 = <LazyCell<i32>>::new(None);
+    assert_eq!(
+        format!("{:?}", &c1),
+        "LazyCell { key: None, cache: Some(Entry { value: None, state: Mutated }) }",
+    );
+    let c2 = <LazyCell<i32>>::new(Some(42));
+    assert_eq!(
+        format!("{:?}", &c2),
+        "LazyCell { key: None, cache: Some(Entry { value: Some(42), state: Mutated }) }",
+    );
+    let c3 = <LazyCell<i32>>::lazy(Key([0x00; 32]));
+    assert_eq!(
+        format!("{:?}", &c3),
+        "LazyCell { \
+            key: Some(Key(\
+                0x00_\
+                00000000_00000000_\
+                00000000_00000000_\
+                00000000_00000000_\
+                00000000_000000)), \
+            cache: None \
+        }",
+    );
+}
+
+impl<T> Drop for LazyCell<T>
+where
+    T: SpreadLayout,
+{
+    fn drop(&mut self) {
+        if let Some(key) = self.key() {
+            if let Some(entry) = self.entry() {
+                if <T as SpreadLayout>::REQUIRES_DEEP_CLEAN_UP {
+                    // We need to load the entity before we remove its associated contract storage
+                    // because it requires a deep clean-up which propagates clearing to its fields,
+                    // for example in the case of `T` being a `storage::Box`.
+                    clear_spread_root_opt::<T>(entry.value().into(), key)
+                } else {
+                    // The type does not require deep clean-up so we can simply clean-up
+                    // its associated storage cell and be done without having to load it first.
+                    let footprint = <T as SpreadLayout>::FOOTPRINT;
+                    assert!(
+                        footprint <= 32,
```

Comment:
Similar to fragment in `optspec::clear_associated_storage_cells` except that is `footprint < 32` is that intentional? Also should this limit be documented for the `FOOTPRINT` trait constant?

Reply:
good found! I wanted to refactor clean up of storage anyways so I now just did. Resolved.